core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -194,19 +205,23 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
    */
   final def flatMap[R1 <: R, E1 >: E, B](f0: A => ZManaged[R1, E1, B]): ZManaged[R1, E1, B] =
     ZManaged[R1, E1, B] {
-      Ref.make[ZIO[R1, Nothing, Any]](IO.unit).map { finalizers =>
+      Ref.make[List[ZIO[R1, Nothing, Any]]](Nil).map { finalizers =>
         Reservation(
           acquire = for {
             resR <- reserve
-                     .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                     .flatMap(res => finalizers.update(res.release :: _).const(res))
                      .uninterruptible
             r <- resR.acquire
             resR1 <- f0(r).reserve
-                      .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                      .flatMap(res => finalizers.update(res.release :: _).const(res))
                       .uninterruptible
             r1 <- resR1.acquire
           } yield r1,
-          release = ZIO.flatten(finalizers.get)
+          release = for {
+            fs    <- finalizers.get
+            exits <- ZIO.foreach(fs)(_.run)
+            _     <- ZIO.done(Exit.all(exits).getOrElse(Exit.unit))
```

Comment:
Since you're not using the result anyway:
```suggestion
            _     <- ZIO.done(Exit.all(exits))
```